File: Java/LinkedLists/Remove Loop.java
package LinkedLists;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

class Node<T> {
    private T data;
    private Node<T> next;

    public Node(T data) {
        this.data = data;
        this.next = null;
    }

    public Node<T> getNext() {
        return next;
    }

    public void setNext(Node<T> next) {
        this.next = next;
    }

    public T getData() {
        return data;
    }
}

class LinkedList<T> {
    private Node<T> head;
    private Node<T> tail;
    private Integer length;

    public Node<T> getHead() {
        return head;
    }

    public void setHead(Node<T> head) {
        this.head = head;
    }

    public Node<T> getTail() {
        return tail;
    }

    public void setTail(Node<T> tail) {
        this.tail = tail;
    }

    public LinkedList() {
        this.head = this.tail = null;
        this.length = 0;
    }

    public boolean isEmpty() {
        return this.length.equals(0);
    }

    public void push(T x) {
        Node<T> node = new Node<>(x);
        if (isEmpty()) {
            setHead(node);
            setTail(node);
        }
        else {
            tail.setNext(node);
            setTail(node);
        }
        length += 1;
    }

    @Override
    public String toString() {
        if (isEmpty()) {
            return "[]";
        }
        StringBuilder result = new StringBuilder(String.format("[%s", getHead().getData()));
        Node<T> curr = getHead().getNext();
        while (curr != getTail()) {
            result.append(String.format(", %s", curr.getData()));
            curr = curr.getNext();
        }
        result.append(String.format(", %s]", getTail().getData()));
        return result.toString();
    }

    public boolean hasLoop() {
        Node<T> slow = getHead();
        Node<T> fast = getHead();
        while (slow != null && fast != null && fast.getNext() != null) {
            slow = slow.getNext();
            fast = fast.getNext().getNext();
            if(slow == fast) {
                return true;
            }
        }
        return false;
    }

    public void removeLoop() {
        if (hasLoop()) {
            getTail().setNext(null);
        }
    }
}

class Solution {
    public static void main(String[] args) {
        List<Integer> list = Arrays.asList(1, 2, 3, 4, 10, -19);
        LinkedList<Integer> linkedList = new LinkedList<>();
        list.forEach(linkedList::push);
        System.out.println(linkedList);
        System.out.println(linkedList.hasLoop());
        linkedList.getTail().setNext(linkedList.getHead().getNext().getNext());
        System.out.println(linkedList.hasLoop());
        linkedList.removeLoop();
        System.out.println(linkedList.hasLoop());
    }
}

File: Python/Practice Set 2/Linked List/remove loop.py
class Node:
    def __init__(self, data):
        self.data = data
        self.next = None


class LinkedList:
    def __init__(self):
        self.head = self.tail = None
        self.length = 0

    def is_empty(self):
        return self.length == 0

    def push(self, x):
        node = Node(x)
        if self.is_empty():
            self.head = self.tail = node
        else:
            self.tail.next = node
            self.tail = node
        self.length += 1

    def show(self):
        curr = self.head
        while curr is not None:
            print(curr.data, end=" ")
            curr = curr.next
        print()

    def build(self, *args):
        for i in args:
            self.push(i)

    def has_loop(self):
        slow = fast = self.head
        while slow and fast and fast.next:
            slow = slow.next
            fast = fast.next.next
            if slow == fast:
                return True
        return False

    def remove_loop(self):
        self.tail.next = None


# Example 1
l1 = LinkedList()
l1.build(1, 2, 3, 4, 5)
l1.show()
print(l1.has_loop())
l1.tail.next = l1.head.next
print(l1.has_loop())
l1.remove_loop()
print(l1.has_loop())
